### PR TITLE
Admin docs inlining part 2

### PR DESCRIPTION
**Pull request by iainmcampbell**
Part of https://github.com/Shopify/argo-private/issues/1507

export shared types: since the docs generator script starts parsing component types from `src/components/index.ts`, this puts these shared types where it can access them.

inline API docs: same as we did for components (https://github.com/Shopify/argo-checkout/pull/82)

### CODE REVIEW

File: packages/argo-admin-react/src/extension-api/data/examples/Data.example.tsx [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+      - variant id: ${variantId}
+    `}
+    </Text>
+  );
+}
+
+extend(
+  'Admin::Product::SubscriptionPlan::Add',
+  render(() => <App />),
+);
```
Comment (elanalynn) [inline packages/argo-admin-react/src/extension-api/data/examples/Data.example.tsx:21]: It may worth adding the EOF new lines to all these new files so they don't get modified with unrelated code changes later.

**Review by elanalynn — APPROVED** [review]
This is great (and he build works 🎩)!

**Review by Fionoble — APPROVED** [review]
Nice 👍

File: packages/argo-admin-react/src/extension-api/data/examples/data.example.tsx [hunk 2]
```diff
@@ -0,0 +1,21 @@
+import React from 'react';
+import {extend, render, Text, useData} from '@shopify/argo-admin-react';
+
+function App() {
+  const {productId, variantId} = useData();
```
Comment (vividviolet) [inline packages/argo-admin-react/src/extension-api/data/examples/data.example.tsx:5]: We should be passing the extension point here for type safety
`const {productId, variantId} = useData<Admin::Product::SubscriptionPlan::Add>();`

File: packages/argo-admin/src/extension-api/container/index.ts [hunk 3]
```diff
@@ -1,9 +1,12 @@
-import {ExtensionPoint, ExtensionApi} from '../extension-points';
+import {ExtensionPoint, ExtensionApi} from '../../extension-points';
```
Comment (vividviolet) [inline packages/argo-admin/src/extension-api/container/index.ts:1]: No docs for this?

**Review by vividviolet — COMMENTED** [review]
Can we add docs for `useExtensionApi`? It's currently inside `utils`

File: packages/argo-admin-react/src/extension-api/data/examples/data.example.tsx (hunk 2 above)
Comment (iainmcampbell) [reply packages/argo-admin-react/src/extension-api/data/examples/data.example.tsx:5]: so far, I've been writing the examples using typescript but not actually doing any type annotations. this way, vscode catches obvious issues (like the fact that we had extensionPoint imported for every example) but the examples show up in 'JS'. probably should just have tsx and jsx versions for each though…

File: packages/argo-admin/src/extension-api/container/index.ts (hunk 3 above)
Comment (iainmcampbell) [reply packages/argo-admin/src/extension-api/container/index.ts:1]: I'm still thinking about how to document extension points. They're not near as simple as component, or even APIs, for which listing the props + a brief intro pretty much covers.

"listing the props" for an extension point mean something more like a list of allowable components and APIs, plus the actual props for the data API. however, the way we've implemented extension points is quite complicated to parse with tsdoc. I'm treating it as a separate task, and thinking about how automated it should really be.

Comment (iainmcampbell) [reply packages/argo-admin/src/extension-api/container/index.ts:1]: tldr this is my next task after the other APIs

File: packages/argo-admin-react/src/extension-api/data/examples/data.example.tsx (hunk 2 above)
Comment (vividviolet) [reply packages/argo-admin-react/src/extension-api/data/examples/data.example.tsx:5]: For the `useData` and `useContainer` hooks you need to explicitly pass in the extension point otherwise the data object comes back as `any`. I don't think we want to encourage this? We want people to be able to get type safety for the data.

File: packages/argo-admin/src/extension-api/container/index.ts (hunk 3 above)
Comment (iainmcampbell) [reply packages/argo-admin/src/extension-api/container/index.ts:1]: same problem as `data` - the way we've typed the containers using generics on the extension point is quite hard to parse with tsdoc. there's like 3 levels of tracing generics to get from this file to

```
data: {
  productId: string;
  variantId?: string;
};
```

File: packages/argo-admin-react/src/extension-api/data/examples/data.example.tsx (hunk 2 above)
Comment (iainmcampbell) [reply packages/argo-admin-react/src/extension-api/data/examples/data.example.tsx:5]: yeah, we should just have JS and TS examples. i'll add them.

Can you get type safety using the typescript vanilla API?

File: packages/argo-admin/src/extension-api/container/index.ts (hunk 3 above)
Comment (iainmcampbell) [reply packages/argo-admin/src/extension-api/container/index.ts:1]: we also didn't have an example for useContainer in the existing docs: https://github.com/Shopify/argo-admin-template/blob/master/docs/Utilities/Container.md

File: packages/argo-admin-react/src/extension-api/data/examples/data.example.tsx (hunk 2 above)
Comment (vividviolet) [reply packages/argo-admin-react/src/extension-api/data/examples/data.example.tsx:5]: @iainmcampbell you get it for free in vanilla without doing anything. Only the hooks require manually specifying the extension point.

image

File: packages/argo-admin/src/extension-api/container/index.ts (hunk 3 above)
Comment (vividviolet) [reply packages/argo-admin/src/extension-api/container/index.ts:1]: Maybe we should remove the docs for `useData` to be consistent then?

File: packages/argo-admin-react/src/extension-api/data/examples/data.example.tsx (hunk 2 above)
Comment (iainmcampbell) [reply packages/argo-admin-react/src/extension-api/data/examples/data.example.tsx:5]: ok so the way we have it in the template (define a named function, then pass that to `extend()` does require you to manually specify the extension point:

Screen Shot 2021-05-05 at 1 57 11 PM

Screen Shot 2021-05-05 at 1 57 23 PM

but if you pass an anonymous function to `extend()` you do get the types for free.
Screen Shot 2021-05-05 at 1 57 58 PM

File: packages/argo-admin/src/extension-api/container/index.ts (hunk 3 above)
Comment (iainmcampbell) [reply packages/argo-admin/src/extension-api/container/index.ts:1]: i'm gonna write an example, better to have than not.

**Comment by iainmcampbell** [comment]
@vividviolet added examples for data and container

**Review by vividviolet — APPROVED** [review]
Looks great! 👏
